crates/oxc_minifier/src/printer/gen.rs

```diff
@@ -793,23 +793,88 @@ impl Gen for NullLiteral {
 impl<'a> Gen for NumberLiteral<'a> {
     #[allow(clippy::cast_sign_loss, clippy::cast_possible_truncation)]
     fn gen(&self, p: &mut Printer) {
-        if self.base != NumberBase::Float {
+        p.print_space_before_identifier();
+
+        let result = if self.base == NumberBase::Float {
+            print_non_negative_float(self.value, p)
+        } else {
             let value = self.value as u64;
-            let decimal = format!("{value}");
-            let hex = format!("{value:#x}");
-            let s = if (1_000_000_000_000..=0xFFFF_FFFF_FFFF_F800).contains(&value) {
+            // If integers less than 1000, we know that exponential notation will always be longer than
+            // the integer representation. This is not the case for 1000 which is "1e3".
+            if value < 1000 {
+                format!("{value}")
+            } else if (1_000_000_000_000..=0xFFFF_FFFF_FFFF_F800).contains(&value) {
                 let hex = format!("{value:#x}");
-                if hex.len() < decimal.len() { hex } else { decimal }
+                let result = print_non_negative_float(self.value, p);
+                if hex.len() < result.len() { hex } else { result }
             } else {
-                decimal
-            };
-            p.print_str(s.as_bytes());
-            return;
+                print_non_negative_float(self.value, p)
+            }
+        };
+        let bytes = result.as_bytes();
+        p.print_str(bytes);
+
+        // We'll need a space before "." if it could be parsed as a decimal point
+        if !bytes.iter().any(|&b| matches!(b, b'.' | b'e' | b'x')) {
+            p.need_space_before_dot = p.code_len();
         }
-        p.print_str(self.raw.as_bytes());
     }
 }
 
+fn print_non_negative_float(value: f64, p: &mut Printer) -> String {
```

Comment:
The algorithm works, but I think we should be able to refactor this into the least number of allocations. A small task @anonrig may be able to help in the future.

Reply:
Can you add a `TODO: refactor this with less allocations` comment?